dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/metrics/BatchMetricsDispatcher.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.metrics
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.api.feature.Feature
+import com.datadog.android.core.internal.configuration.DataUploadConfiguration
+import com.datadog.android.core.internal.persistence.file.FilePersistenceConfig
+import com.datadog.android.core.internal.time.TimeProvider
+import com.datadog.android.core.sampling.RateBasedSampler
+import java.io.File
+
+internal class BatchMetricsDispatcher(
+    featureName: String,
+    private val uploadConfiguration: DataUploadConfiguration,
+    private val filePersistenceConfig: FilePersistenceConfig,
+    private val internalLogger: InternalLogger,
+    private val dateTimeProvider: TimeProvider,
+    private val sampler: RateBasedSampler = RateBasedSampler(15f)
+) :
+    MetricsDispatcher {
+
+    private val trackName: String? = resolveTrackName(featureName)
+
+    override fun sendBatchDeletedMetric(batchFile: File, deletionReason: RemovalReason) {
+        if (!deletionReason.includeInMetrics()) {
+            return
+        }
+
+        if (trackName == null) {
+            return
+        }
+
+        if (!sampler.sample()) {
+            return
+        }
+        val batchDeletedMetricAttributes =
+            resolveBatchDeletedMetricAttributes(batchFile, deletionReason)
+        internalLogger.metric(
+            messageBuilder = { BATCH_DELETED_MESSAGE },
+            additionalProperties = batchDeletedMetricAttributes
+        )
+    }
+
+    private fun resolveBatchDeletedMetricAttributes(file: File, deletionReason: RemovalReason): Map<String, Any?> {
+        val fileAgeInMillis = resolveFileAge(file)
+        return mutableMapOf<String, Any?>(
+            TRACK_KEY to trackName,
+            TYPE_KEY to BATCH_DELETED_TYPE_VALUE,
+            BATCH_AGE_KEY to fileAgeInMillis,
+            UPLOADER_DELAY_KEY to mapOf(
+                UPLOADER_DELAY_MIN_KEY to uploadConfiguration.minDelayMs,
+                UPLOADER_DELAY_MAX_KEY to uploadConfiguration.maxDelayMs
+            ),
+            UPLOADER_WINDOW_KEY to filePersistenceConfig.recentDelayMs,
+
+            BATCH_REMOVAL_KEY to deletionReason.toString(),
+            IN_BACKGROUND_KEY to false
+        )
+    }
+
+    private fun resolveFileAge(file: File): Long {
+        return dateTimeProvider.getServerTimestamp() - file.asTimestamp()
```

Comment:
question/ 💡 Do both `getServerTimestamp()` and `file.asTimestamp()` use the same time source? The first sounds to be NTP-fixed server time vs the other is likely a device time. If that's the case, this could go even negative.

Reply:
oh that's a good catch...was actually counting on both using the same time provider. Need to check that.